src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -323,3 +498,236 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
+                                         "**opnotpredefined", "**opnotpredefined %d", op);
+        return mpi_errno;
+        /* --END ERROR HANDLING-- */
+    }
+
+    if (!(is_empty_source == TRUE || HANDLE_IS_BUILTIN(target_dtp))) {
+        MPIR_Datatype_get_ptr(target_dtp, dtp);
+        MPIR_Assert(dtp != NULL);
+        vec_len = dtp->typerep.num_contig_blocks * target_count + 1;
+        /* +1 needed because Rob says so */
+        typerep_vec = (struct iovec *)
+            MPL_malloc(vec_len * sizeof(struct iovec), MPL_MEM_OTHER);
+        /* --BEGIN ERROR HANDLING-- */
+        if (!typerep_vec) {
+            mpi_errno =
+                MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
+                                     MPI_ERR_OTHER, "**nomem", 0);
+            goto fn_exit;
+        }
+        /* --END ERROR HANDLING-- */
+
+        MPI_Aint actual_iov_len, actual_iov_bytes;
+        MPIR_Typerep_to_iov(NULL, target_count, target_dtp, 0, typerep_vec, vec_len,
+                            source_count * source_dtp_size, &actual_iov_len, &actual_iov_bytes);
+        vec_len = actual_iov_len;
+
+        type = dtp->basic_type;
+        MPIR_Assert(type != MPI_DATATYPE_NULL);
+
+        MPIR_Assert(type == source_dtp);
+        type_size = source_dtp_size;
+        type_extent = source_dtp_extent;
+        /* If the source buffer has been packed by the caller, the distance between
+         * two elements can be smaller than extent. E.g., predefined pairtype may
+         * have larger extent than size.*/
+        /* when predefined pairtype have larger extent than size, we'll end up
+         * missaligned access. Memcpy the source to workaround the alignment issue.
+         */
+        if (src_kind == MPIDIG_ACC_SRCBUF_PACKED) {
+            src_type_stride = source_dtp_size;
+            if (source_dtp_size < source_dtp_extent) {
+                src_ptr = MPL_malloc(source_dtp_extent, MPL_MEM_OTHER);
+            }
+        } else {
+            src_type_stride = source_dtp_extent;
+        }
+    }
+    if (is_supported) {
+        if (is_empty_source == TRUE || HANDLE_IS_BUILTIN(target_dtp)) {
+            /* directly apply op if target dtp is predefined dtp OR source buffer is empty */
+            max_pack_bytes = source_dtp_size * source_count;
+            mpi_errno =
+                op_pack(source_buf, source_count, source_dtp, 0, target_buf,
+                        max_pack_bytes, &actual_pack_bytes, op);
+
+        } else {
+            /* derived datatype */
+            i = 0;
+            curr_loc = typerep_vec[0].iov_base;
+            curr_len = typerep_vec[0].iov_len;
+            accumulated_count = 0;
+            while (i != vec_len) {
+                if (curr_len < type_size) {
+                    MPIR_Assert(i != vec_len);
+                    i++;
+                    curr_len += typerep_vec[i].iov_len;
+                    continue;
```

Comment:
I don't quite understand this code

Reply:
This code is same for both yaksa and host to go over the iovs.

Reply:
I was wondering how the `curr_len` is omitting the gap between iovs. It appears the code is omitting the gap because it is calculating count by `datatype_size`. Anyway, the code was too subtle to read.